dingo/core/utils/pt_to_hdf5.py

```diff
@@ -0,0 +1,77 @@
+        description="Convert the weights of a trained PyTorch model to HDF5 format"
+        " for distribution in the LVK's CVMFS.",
+        epilog="Training history (optimizer_state_dict) is discarded.")
+    parser.add_argument("-i", "--in_file", type=str, required=True,
+            help='Input model ".pt" weights file')
+    parser.add_argument("-o", "--out_file", type=str, required=True,
+            help='Output model ".hdf5" weights file')
+    parser.add_argument("-n", "--version_number", type=int, required=True,
+            help="Model version number (integer). "
+            "Will be included in the output filename and metadata.")
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    if os.path.splitext(args.in_file)[-1] != '.pt':
+        raise ValueError('Expected a .pt input file')
+    if os.path.splitext(args.out_file)[-1] != '.hdf5':
+        raise ValueError('Expected a .hdf5 output file')
+
+    # Build output filename with version number
+    # This is required for use on CVMFS
+    version_number = args.version_number
+    root, ext = os.path.splitext(args.out_file)
+    out_file_name = f'{root}_v{version_number}{ext}'
+    print('Output will be written to', out_file_name)
+
+    # Load data into CPU memory since we'll be saving it using CPU libraries
+    d = torch.load(args.in_file, map_location=torch.device("cpu"))
+
+    # Collect the names of the dicts that can be serialized to JSON; model_state_dict and
+    # optimizer_state_dict contain torch.tensors and cannot be JSON serialized
+    dicts_to_serialize = ['model_kwargs', 'epoch', 'metadata', 'optimizer_kwargs',
+            'scheduler_kwargs', 'scheduler_state_dict']
+
+
+    with h5py.File(out_file_name, 'w') as f:
+        # Save small nested dicts as json
+        grp = f.create_group('serialized_dicts')
+        for k in dicts_to_serialize:
+            dict_str = json.dumps(d[k])
+            grp.create_dataset(k, data=dict_str)
+
+        # Save the OrderedDict containing the model weights
+        # The keys are ordered alphanumerically as well.
+        grp_model = f.create_group('model_weights')
+        for k, v in d['model_state_dict'].items():
+            if len(v.size()) > 0:
+                grp_model.create_dataset(k, data=v.numpy(), fletcher32=True)
+            else:
+                # fletcher32 is not available for scalars
+                grp_model.create_dataset(k, data=v.numpy())
+
+        # Note we do not save optimizer_state_dict which is
+        # not needed at inference time and saves a significant
+        # amount of disk space.
+
+        # Metadata for CVMFS LVK distribution
+        # This needs to be exactly the same as the "basename" of the hdf5 file
+        f.attrs['CANONICAL_FILE_BASENAME'] = os.path.basename(out_file_name)
+        # Add a few metadata entries as attributes
+        f.attrs['approximant'] = d['metadata']['dataset_settings']['waveform_generator']['approximant']
```

Comment:
What is the reason for this? To just have easier access to the approximant?

Reply:
Yes, this is just an example. I think it is useful to have it top-level, but we might also encode it in the model filename.